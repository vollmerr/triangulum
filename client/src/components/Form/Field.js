import React from 'react';
import { Form } from 'semantic-ui-react';

// maps formik props to semantic-ui field
const Field = ({ formikProps, field }) => {
  const {
    values,
    touched,
    errors,
    handleChange,
    handleBlur,
    setFieldValue,
    setFieldTouched,
  } = formikProps;
  const { name, options } = field;

  const props = {
    ...field,
    value: values[name],
    onChange: handleChange,
    onBlur: handleBlur,
    error: !!errors[name] && touched[name],
  };

  // handle fields with multiple values (select, etc)
  if (options) {
    props.options = options.map(option => ({
      name,
      key: option.value,
      ...option,
    }));
    props.onChange = (e, x) => setFieldValue(name, x.value);
    props.onBlur = () => setFieldTouched(name, true);
  }

  return (
    <Form.Field {...props} />
  );
};

export default Field;
