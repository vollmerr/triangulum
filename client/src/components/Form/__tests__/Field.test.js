import React from 'react';
import { shallow } from 'enzyme';

import Field from '../Field';

const formikProps = {
  values: {
    k1: 'v1',
  },
  touched: {},
  errors: {},
  handleChange: jest.fn(),
  handleBlur: jest.fn(),
  setFieldValue: jest.fn(),
  setFieldTouched: jest.fn(),
  isSubmitting: false,
};

const field = {
  name: 'k1',
  options: [
    { value: 'o1', text: 'option 1' },
    { value: 'o2', text: 'option 2' },
  ],
};

const props = {
  formikProps,
  field,
};

let wrapper;
let FormField;
describe('FormikForm', () => {
  beforeEach(() => {
    wrapper = shallow(<Field {...props} />);
    FormField = wrapper.first();
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass an error if error is set and touched', () => {
    wrapper.setProps({
      formikProps: {
        ...props.formikProps,
        errors: { k1: 'k1 error' },
        touched: { k1: true },
      },
    });
    const error = wrapper.first().prop('error');
    expect(error).toBeTruthy();
  });

  describe('single value fields', () => {
    beforeEach(() => {
      wrapper.setProps({ field: { ...field, options: null } });
      FormField = wrapper.first();
    });

    it('should map the `onChange` correctly', () => {
      FormField.prop('onChange')('v2');
      expect(formikProps.handleChange).toBeCalledWith('v2');
    });

    it('should map the `onBlur` correctly', () => {
      FormField.prop('onBlur')();
      expect(formikProps.handleBlur).toBeCalled();
    });
  });

  describe('multi value fields', () => {
    it('should map the options correctly', () => {
      const options = FormField.prop('options');
      expect(options.length).toBe(field.options.length);
      expect(options[0].name).toBe(field.name);
      expect(options[0].value).toBe(field.options[0].value);
    });

    it('should map the `onChange` correctly', () => {
      const option = field.options[1];
      FormField.prop('onChange')(null, option);
      expect(formikProps.setFieldValue).toBeCalledWith(field.name, option.value);
    });

    it('should map the `onBlur` correctly', () => {
      FormField.prop('onBlur')();
      expect(formikProps.setFieldTouched).toBeCalledWith(field.name, true);
    });
  });
});
