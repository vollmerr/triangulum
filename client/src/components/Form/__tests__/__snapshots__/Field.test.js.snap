// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormikForm should render correctly 1`] = `
<FormField
  data-cy="field-test"
  disabled={false}
  error={false}
  name="k1"
  onBlur={[Function]}
  onChange={[Function]}
  options={
    Array [
      Object {
        "key": "o1",
        "name": "k1",
        "text": "option 1",
        "value": "o1",
      },
      Object {
        "key": "o2",
        "name": "k1",
        "text": "option 2",
        "value": "o2",
      },
    ]
  }
  value="v1"
/>
`;
